### PR TITLE
feat: ovn eip snat fip

### PR DESCRIPTION
**Pull request by bobz965**
- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

support vpc use ovn eip fip snat  by crd , compatiblewith  provider network, vlan, subnet, compatible with previous default vpc  enable_eip_snat

#### What type of this PR
- Features

### CODE REVIEW

**Comment by bobz965** [comment]
[10-support-ovn-eip-fip-snat-draft.md](https://github.com/kubeovn/kube-ovn/files/9971076/10-support-ovn-eip-fip-snat-draft.md)

**Comment by bobz965** [comment]
test previous ovn-default default vpc enable eip fip works well and not changed:
[01-test-kube-ovn-eip-snat-ovn-default-vpc-as-previous.txt](https://github.com/kubeovn/kube-ovn/files/9979535/01-test-kube-ovn-eip-snat-ovn-default-vpc-as-previous.txt)

File: pkg/controller/ovn_eip.go [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+	}
+	portName := cachedEip.Name
+	if cachedEip.Spec.V4Ip != "" {
+		v4ip, v6ip, mac, err = c.acquireStaticIpAddress(subnet.Name, cachedEip.Name, portName, cachedEip.Spec.V4Ip)
+	} else {
+		// random allocate
+		v4ip, v6ip, mac, err = c.acquireIpAddress(subnet.Name, cachedEip.Name, portName)
+	}
+	if err != nil {
+		return err
```
Comment (hongzhen-ma) [inline pkg/controller/ovn_eip.go:244]: 加个log比较好

File: pkg/controller/ovn_eip.go [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	if cachedEip.Status.MacAddress != "" && cachedEip.Status.MacAddress != cachedEip.Spec.MacAddress {
+		// eip not support change ip, reset eip spec from its status
+		if err = c.resetOvnEipSpec(key); err != nil {
+			return err
```
Comment (hongzhen-ma) [inline pkg/controller/ovn_eip.go:295]: 这也加个log吧

File: pkg/controller/vpc.go [hunk 3]
```diff
@@ -110,6 +110,10 @@ func (c *Controller) handleDelVpc(vpc *kubeovnv1.Vpc) error {
 	if err != nil {
 		return err
 	}
+
+	if err := c.handleDelVpcExternal(vpc.Name); err != nil {
```
Comment (hongzhen-ma) [inline pkg/controller/vpc.go:114]: 也加个log

File: pkg/controller/vpc.go [hunk 4]
```diff
@@ -720,3 +738,103 @@ func (c *Controller) createVpcRouter(lr string) error {
...
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return err
+		}
+		needCreateEip = true
+	}
+	var v4ip, v6ip, mac string
+	klog.V(3).Infof("create vpc lrp eip %s", lrpEipName)
+	if needCreateEip {
+		if v4ip, v6ip, mac, err = c.acquireIpAddress(c.config.ExternalGatewaySwitch, lrpEipName, lrpEipName); err != nil {
```
Comment (hongzhen-ma) [inline pkg/controller/vpc.go:759]: 申请地址失败，也加个log吧

File: pkg/controller/vpc.go [hunk 5]
```diff
@@ -720,3 +738,103 @@ func (c *Controller) createVpcRouter(lr string) error {
...
+		klog.Errorf("failed to get ovn-external-gw-config, %v", err)
+		return err
+	}
+	chassises, err := c.getGatewayChassis(cm.Data)
+	if err != nil {
+		klog.Errorf("failed to get gateway chassis, %v", err)
+		return err
+	}
+	v4ipCidr := util.GetIpAddrWithMask(v4ip, cachedSubnet.Spec.CIDRBlock)
+	if err := c.ovnLegacyClient.ConnectRouterToExternal(c.config.ExternalGatewaySwitch, key, v4ipCidr, mac, chassises); err != nil {
```
Comment (hongzhen-ma) [inline pkg/controller/vpc.go:788]: 关键配置报错，也加个log吧

File: pkg/controller/vpc.go [hunk 6]
```diff
@@ -720,3 +738,103 @@ func (c *Controller) createVpcRouter(lr string) error {
...
+}
+
+func (c *Controller) handleDelVpcExternal(key string) error {
+	lrpEipName := fmt.Sprintf("%s-%s", key, c.config.ExternalGatewaySwitch)
+	cachedEip, err := c.ovnEipsLister.Get(lrpEipName)
+	if err != nil {
+		return err
+	}
+	klog.V(3).Infof("delete vpc lrp %s", lrpEipName)
+	if err := c.ovnLegacyClient.DisconnectRouterToExternal(c.config.ExternalGatewaySwitch, key); err != nil {
```
Comment (hongzhen-ma) [inline pkg/controller/vpc.go:826]: 这个地方也加一下log

**Review by hongzhen-ma — CHANGES_REQUESTED** [review]

**Pull request merged**